docs/glossary.md

```diff
@@ -0,0 +1,211 @@
+
+A snapshot of the files in the repository at a given point in time (technically
+a [tree object](#tree)), together with some metadata. The metadata includes the
+author, the date, and pointers to the commit's parents. Through the pointers to
+the parents, the commits form a
+[Directed Acyclic Graph (DAG)](https://en.wikipedia.org/wiki/Directed_acyclic_graph)
+.
+
+Note that even though commits are stored as snapshots, they are often treated
+as differences between snapshots, namely compared to their parent's snapshot. If
+they have more than one parent, then the difference is computed against the
+result of merging the parents. For example, `jj diff` will show the differences
+introduced by a commit compared to its parent(s), and `jj rebase` will apply
+those changes onto another base commit.
+
+The word "revision" is used as a synonym for "commit".
+
+## Commit ID
+
+A commit ID is a unique identifier for a [commit](#commit). They are 20 bytes
+long when using the Git backend. They are presented in regular hexadecimal
+format at the end of the line in `jj log`, using 12 hexadecimal digits by
+default. When using the Git backend, the commit ID is the Git commit ID.
+
+## Co-located repos
+
+When using the Git [backend](#backend) and the backing Git repository's `.git/`
+directory is a sibling of `.jj/`, we call the repository "co-located". Most
+tools designed for Git can be easily used on such repositories. `jj` and `git`
+commands can be used interchangeably.
+
+See [here](git-compatibility.md#co-located-jujutsugit-repos) for details.
+
+## Conflict
+
+Conflicts can occur in many places. The most common type is conflicts in files.
+Those are the conflicts that users coming from other VCSs are usually familiar
+with. You can see them in `jj status` and in `jj log` (the red "conflict"
+label at the end of the line). See [here](conflicts.md) for details.
+
+Conflicts can also occur in [branches](#branch). For example, if you moved a
+branch locally, and it was also moved on the remote, then the branch will be
+in a conflicted state after you pull from the remote.
+See [here](branches.md#conflicts) for details.
+
+Similar to a branch conflict, when a [change](#change) is rewritten locally
+and remotely, for example, then the change will be in a conflicted state. We
+call that a [divergent change](#divergent-change).
+
+## Divergent change
+
+A divergent change is a [change](#change) that has more than one
+[visible commit](#visible-commits).
+
+## Head
+
+A head is a commit with no descendants. The context in which it has no
+descendants varies. For example, the `heads(X)`
+[revset function](revsets.md#functions) returns commits that have no descendants
+within the set `X` itself. The [view](#view) records which
+anonymous heads (heads without a branch pointing to them) are visible at a
+given [operation](#operation). Note that this is quite different from Git's
+[HEAD](https://git-scm.com/book/en/v2/Git-Internals-Git-References#ref_the_ref).
```

Comment:
(No action required)
This section is really well-worded :)

Reply:
I disagree that this is well-worded, because it mixes several things: parentage (by mentioning descendants), revsets, view, visibility, and the graph(or tree by mentioning branches).
It sounds to me like the DAG is referred to by tree terms, so why call it a `head` when it's a `leaf` in the tree? Only because Git does? (even though you say it's quite different from Git)

Reply:
I think you're both right :). This section is clear enough to show problems with the terminology we're using.

I also find the "anonymous head" terminology confusing, since 1) it's actually about visibility of commits and 2) if I understand correctly, it doesn't even have to be anonymous.

On that note, if I don't forget, I'll send in a correction to what seems to me to be a minor mistake here. Where it says "heads without a branch pointing to them", I think it should say "heads that may or may not have a named branch pointing to them".

I think it would be nice to come up with a better term. One possibility that comes to my mind is "visible head". "visible leaf" also would seem to work, but then we'd need to replace "head" with "leaf" throughout the code.

Reply:
I say there would be less confusion if you don't use the same term as Git, with a different meaning.